### PR TITLE
Add WithServerName ClientOption

### PR DESCRIPTION
**Pull request by APShirley**
This helps us implement mutual TLS.

### CODE REVIEW

**Comment by APShirley** [comment]
Re-opened to update that the CLA has been signed by both of us.

File: config_test.go [hunk 1]
```diff
@@ -23,6 +23,8 @@ type buildResult struct {
 }
 
 func TestE2E(t *testing.T) {
+	t.Parallel()
```
Comment (sunjayBhatia) [inline config_test.go]: I believe that we had placed these later b/c the `certstrap` library the `certtest` package uses is not safe for concurrent use: https://github.com/square/certstrap/issues/72

Comment (sunjayBhatia) [reply config_test.go]: We might end up seeing flakiness in CI, with this specific change, we can validate by bumping it in diego-release

Comment (sunjayBhatia) [reply config_test.go]: see: https://github.com/cloudfoundry/tlsconfig/commit/d8e16d7d2c43cc9ebe0d0c15679ecc9c948a7912

Comment (xoebus) [reply config_test.go]: Ah, I’m sorry. I’ll revert this bit and tidy up this afternoon.

Comment (APShirley) [reply config_test.go]: Hi @sunjayBhatia @xoebus, I figured this could possibly be problematic, hence it was added as a separate commit. Revert away!

**Pull request merged**